Python/specialize.c

```diff
@@ -1957,6 +2008,93 @@ _Py_Specialize_CompareOp(PyObject *lhs, PyObject *rhs, _Py_CODEUNIT *instr,
     cache->counter = initial_counter_value();
 }
 
+char INIT_CLEANUP_CODE[8] = {
+    LOAD_ASSERTION_ERROR, 0,
+    RAISE_VARARGS, 1,
+    EXIT_INIT_CHECK, 0,
+    RETURN_VALUE, 0
+};
+
+PyFunctionObject *_Py_InitCleanupFunc = NULL;
+
+static int
+setup_init_cleanup_func(void) {
+    if (_Py_InitCleanupFunc != NULL) {
+        return 0;
+    }
+    PyObject *name = NULL;
+    PyObject *code = NULL;
+    PyObject *lines = NULL;
+    PyCodeObject *codeobj = NULL;
+    PyObject *globals = NULL;
+
+    name = _PyUnicode_FromASCII("type.__call__", strlen("type.__call__"));
+    if (name == NULL) {
+        goto cleanup;
+    }
+    code = PyBytes_FromStringAndSize(INIT_CLEANUP_CODE, 8);
+    if (code == NULL) {
+        goto cleanup;
+    }
+    const char loc[2] = { 0x80 | (PY_CODE_LOCATION_INFO_NO_COLUMNS << 3) | 3, 0 };
+    lines = PyBytes_FromStringAndSize(loc, 2);
+    if (lines == NULL) {
+        goto cleanup;
+    }
+    struct _PyCodeConstructor con = {
+        .filename = &_Py_STR(empty),
+        .name = name,
+        .qualname = name,
+        .flags = CO_NEWLOCALS | CO_OPTIMIZED,
+
+        .code = code,
+        .firstlineno = 1,
+        .linetable = lines,
+
+        .consts = (PyObject *)&_Py_SINGLETON(tuple_empty),
+        .names = (PyObject *)&_Py_SINGLETON(tuple_empty),
+
+        .localsplusnames = (PyObject *)&_Py_SINGLETON(tuple_empty),
+        .localspluskinds = (PyObject *)&_Py_SINGLETON(bytes_empty),
+
+        .argcount = 0,
+        .posonlyargcount = 0,
+        .kwonlyargcount = 0,
+
+        .stacksize = 2,
+
+        .exceptiontable = (PyObject *)&_Py_SINGLETON(bytes_empty),
+    };
+
+    codeobj = _PyCode_New(&con);
+    if (codeobj == NULL) {
+        goto cleanup;
+    }
+    globals = PyDict_New();
+    if (globals == NULL) {
+        goto cleanup;
+    }
+    PyFrameConstructor desc = {
+        .fc_globals = globals,
+        .fc_builtins = globals,
+        .fc_name = codeobj->co_name,
+        .fc_qualname = codeobj->co_name,
+        .fc_code = (PyObject *)codeobj,
+        .fc_defaults = NULL,
+        .fc_kwdefaults = NULL,
+        .fc_closure = NULL
+    };
+    _Py_InitCleanupFunc = _PyFunction_FromConstructor(&desc);
+cleanup:
+    PyErr_Clear();
+    Py_XDECREF(codeobj);
+    Py_DECREF(globals);
```

Comment:
```suggestion
    Py_XDECREF(globals);
```
This would otherwise segfault if `globals == NULL` because of no memory.